src/editor.c

```diff
@@ -1523,27 +1539,58 @@ static void insert_indent_after_line(GeanyEditor *editor, gint line)
 	g_free(text);
 }
 
+static gboolean auto_close_chars_consume(ScintillaObject *sci, gint pos, gchar c) {
+	gboolean isAutoClosed = FALSE;
+	switch (c)
+	{
+		case ')':
+			if (editor_prefs.autoclose_chars & GEANY_AC_PARENTHESIS)
+				isAutoClosed = TRUE;
+			break;
+		case '}':
+			if (editor_prefs.autoclose_chars & GEANY_AC_CBRACKET)
+				isAutoClosed = TRUE;
+			break;
+		case ']':
+			if (editor_prefs.autoclose_chars & GEANY_AC_SBRACKET)
+				isAutoClosed = TRUE;
+			break;
+		case '\'':
+			if (editor_prefs.autoclose_chars & GEANY_AC_SQUOTE)
+				isAutoClosed = TRUE;
+			break;
+		case '"':
+			if (editor_prefs.autoclose_chars & GEANY_AC_DQUOTE)
+				isAutoClosed = TRUE;
+			break;
+	}
+
+	gchar cNext = sci_get_char_at( sci, pos);
+
+	if( isAutoClosed && cNext == c ) {
+		sci_delete_range(sci, pos, 1);
+		sci_colourise(sci, 0, -1);
+		return TRUE;
+	}
+	return FALSE;
+}
 
 static void auto_close_chars(ScintillaObject *sci, gint pos, gchar c)
 {
 	const gchar *closing_char = NULL;
-	gint end_pos = -1;
-
-	if (utils_isbrace(c, 0))
-		end_pos = sci_find_matching_brace(sci, pos - 1);
```

Comment:
This was deliberately added in https://github.com/geany/geany/commit/ff8664150bab4a9f59120c8d14ca7240ab21a5b6#diff-3fe5ce523684a7fca311fab0d09542552fbe3a28ddaa20ebc5a65c1904bc5e02 so it is desired by at least some people, and nobody (AFAICT) has wanted it changed since 2008.  So removing it just because one person doesn't like it doesn't seem reasonable.

Reply:
Right, but this makes autoclose also next to useless (I have it disabled and would welcome improvements). Since autoclosing now only works if Scintilla doesn't see a matching brace autoclosing does not work most of the time. E.g.:

```
int foo()
{
   if (x) { // !! no autoclose for {, as Scintilla matches the } at the end of the function
}
```
`()` is autoclosed in that example, though. So basically as of now autoclosing works reliably only at the end of documents, otherwise it may seem to work only randomly.